### PR TITLE
Rewrite exercise 4

### LINKED ISSUE #50: Rewrite exercise 4 to version 2
Tasks:
- Rewrite solution to TypeScript
- Switch from reduxsauce to Redux Toolkit
- Switch from `connect` to Hooks API of react-redux
- Update tutorial
- Update slides

### PR DESCRIPTION
**Pull request by dinassh**
## This PR rewrites exercise 4:
- [x] change language to TypeScript
- [x] add smoke tests
- [x] update tutorial for exercise 4
- [x] update slides

## Changes in [slides](https://docs.google.com/presentation/d/1w8HRmAK2HB5PuwOB0HdFeGbvntsvM3-rvHdNOPZ3lBs/edit#slide=id.g8c4a2510bd_3_6):
- pages 33-45: rewritten in TypeScript
- pages 33-36: Redux Toolkit related
- page 38: added mention of Hooks API
- page 43: new page for React Redux Hooks

### Notes
- The code examples font was changed from 11 to 10, because the TypeScript code needs more space than JavaScript and in most cases it wouldn't fit on the page
- I've omitted some includes in code examples due to space issues

### CODE REVIEW

**Review by branoholy — APPROVED** [review]

File: exercises/04-react-redux/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import React from 'react'
+import ReactDOM from 'react-dom'
+
+import { Root } from 'modules/root/components/root'
```
Comment (novellizator) [inline exercises/04-react-redux/src/index.tsx:7]: the import order is a thing of minuscule priority and we don't strictly follow it, but it's good to have on mind that first you import third party libraries and only then you import local stuff

Comment (branoholy) [reply exercises/04-react-redux/src/index.tsx:7]: I think we should add an ESLint rule if we want to follow it.

Comment (dinassh) [reply exercises/04-react-redux/src/index.tsx:7]: Does such rule exist? I see only the ['sort-imports'](https://eslint.org/docs/rules/sort-imports) rule, which 

> verifies that all imports are first sorted by the used member syntax and then alphabetically by the first member or alias name

Comment (branoholy) [reply exercises/04-react-redux/src/index.tsx:7]: Yes, there are these plugins:

- https://github.com/Tibfib/eslint-plugin-import-helpers
- https://github.com/lydell/eslint-plugin-simple-import-sort

Not sure what is exactly different.

Comment (branoholy) [reply exercises/04-react-redux/src/index.tsx:7]: I think the second one can also sort named imports between `{` and `}`. Not sure if we need it. :)

Comment (dinassh) [reply exercises/04-react-redux/src/index.tsx:7]: The change will affect all exercises, so I propose to address it in a separate [issue](https://github.com/salsita/react-training/issues/56)

Comment (novellizator) [reply exercises/04-react-redux/src/index.tsx:7]: Fine with me. Let's just try to follow it going forward ;)

Comment (dinassh) [reply exercises/04-react-redux/src/index.tsx:7]: I've reorganized the imports manually for now. See 0ee4a3c

**Review by novellizator — APPROVED** [review]

**Pull request merged**